File: src/server/api/sync_crdt_handler.go
package api

import (
	"database/sql"
	"encoding/json"
	"net/http"
)

type SyncCRDTHandler struct {
	db *sql.DB
}

func NewSyncCRDTHandler(db *sql.DB) *SyncCRDTHandler {
	return &SyncCRDTHandler{db: db}
}

type CRDTDelta struct {
	ID        string `json:"id"`
	EntityID  string `json:"entity_id"`
	Data      string `json:"data"`
	UpdatedAt string `json:"updated_at"`
}

type CRDTPayload struct {
	Deltas []CRDTDelta `json:"deltas"`
}

func (h *SyncCRDTHandler) HandlePost(w http.ResponseWriter, r *http.Request) {
	var payload CRDTPayload
	if err := json.NewDecoder(r.Body).Decode(&payload); err != nil {
		http.Error(w, "Invalid payload", http.StatusBadRequest)
		return
	}

	tx, err := h.db.Begin()
	if err != nil {
		http.Error(w, "Failed to begin transaction", http.StatusInternalServerError)
		return
	}
	defer tx.Rollback()

	stmt, err := tx.Prepare(`
		INSERT INTO crdt_deltas (id, entity_id, data, updated_at)
		VALUES ($1, $2, $3, $4)
		ON CONFLICT (id) DO UPDATE SET
			data = EXCLUDED.data,
			updated_at = EXCLUDED.updated_at
		WHERE crdt_deltas.updated_at < EXCLUDED.updated_at
	`)
	if err != nil {
		http.Error(w, "Failed to prepare statement", http.StatusInternalServerError)
		return
	}
	defer stmt.Close()

	for _, delta := range payload.Deltas {
		_, err := stmt.Exec(delta.ID, delta.EntityID, delta.Data, delta.UpdatedAt)
		if err != nil {
			http.Error(w, "Failed to insert delta", http.StatusInternalServerError)
			return
		}
	}

	if err := tx.Commit(); err != nil {
		http.Error(w, "Failed to commit transaction", http.StatusInternalServerError)
		return
	}

	w.WriteHeader(http.StatusOK)
}

File: src/server/api/sync_crdt_handler.rs
use axum::{Json, extract::State, response::IntoResponse, http::StatusCode};
use serde::{Deserialize, Serialize};
use sqlx::PgPool;

#[derive(Debug, Deserialize)]
pub struct CrdtDelta {
    pub id: String,
    pub entity_id: String,
    pub data: String,
    pub updated_at: String,
}

#[derive(Debug, Deserialize)]
pub struct CrdtSyncRequest {
    pub deltas: Vec<CrdtDelta>,
}

#[derive(Debug, Serialize)]
pub struct CrdtSyncResponse {
    pub status: String,
}

pub async fn handle_crdt_sync(
    State(pool): State<PgPool>,
    Json(payload): Json<CrdtSyncRequest>,
) -> Result<impl IntoResponse, (StatusCode, String)> {
    let mut tx = pool.begin().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    for delta in payload.deltas {
        let query = r#"
            INSERT INTO crdt_deltas (id, entity_id, data, updated_at)
            VALUES ($1, $2, $3, $4::timestamptz)
            ON CONFLICT (id) DO UPDATE SET
                data = EXCLUDED.data,
                updated_at = EXCLUDED.updated_at
            WHERE crdt_deltas.updated_at < EXCLUDED.updated_at
        "#;

        sqlx::query(query)
            .bind(&delta.id)
            .bind(&delta.entity_id)
            .bind(&delta.data)
            .bind(&delta.updated_at)
            .execute(&mut *tx)
            .await
            .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    }

    tx.commit().await.map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;

    Ok(Json(CrdtSyncResponse {
        status: "success".to_string(),
    }))
}
